### ISSUE TITLE
One database per model?

### ISSUE THREAD

**Issue by perlun** [pouchdb-community/ember-pouch#179]
Hi,

We are looking into using CouchDB for an application of ours. We are caching millions of records from an RDBMS database with some backgrounds jobs. Because it's different kinds of data (customers, addresses, work orders etc) that has its source in different RDBMS tables, we were thinking of using different databases for each type of data.

Like this:

* app_customers
* app_customer_addresses
* app_work_orders
* ...

We will then sync & filter this data into a form suitable for using with CouchDB in offline mode, with PouchDB.

**But** we also want certain parts of the application usable in "online-only" scenarios, going straight towards these databases (or, that was the plan). However, I am not seeing any obvious way to use multiple databases with `ember-pouch`.

This issue discussed it, but no solution is suggested: https://github.com/nolanlawson/ember-pouch/issues/33

Possible workarounds I can think of:

- Using [DS.Store#adapterFor](http://emberjs.com/api/data/classes/DS.Store.html#method_adapterFor) in Ember. Creating one adapter for each model type. However, our app is using perhaps 10-20 different models so it is a bit cumbersome.
- Skip the idea of segregating the data, and live with the fact that we will put perhaps 5 million different records of 10-20 different kinds into a single CouchDB database.

Any better ideas?

**Comment by aexmachina** [comment]
You'll get a new instance of your adapter class for each model class, so you can have a method in the adapter that returns a different db for each model.

**Comment by perlun** [comment]
> You'll get a new instance of your adapter class for each model class, so you can have a method in the adapter that returns a different db for each model.

OK, thanks. What is the _suggested_ approach for how to write an `ember-pouch`-based app, is it best to go with a single CouchDB database or is it better to segregate the data a bit?

/cc @afinne

**Comment by broerse** [comment]
@perlun Also take a look at https://github.com/HospitalRun/hospitalrun-frontend to see their choices, I am myself exploring the option to create a database per users with just the data they need. If each user needs access to the millions of records @aexmachina option can work as long as you don't need all the models every time.

**Comment by afinne** [comment]
What we ended up doing was to have a custom store service, overriding the `adapterFor` method. 

We also have one adapter per database (e.g. one for `customers`, one for `equipment`, one for `appdata`). 

This way we can "point" every model to the correct adapter, and also set up different sync strategies for the different adapters. For instance, customer data and equipment data is read-only for the clients, so we only replicate down (also implementing periodic replication to avoid socket starvation), while appdata is live synced. 

A side effect that we found out after a while, was that the schema building used by relational pouch did not support spreading `belongsTo` and `hasMany` over several adapters, so we moved the schema to the store instead of having it on the adapter.
